### PR TITLE
Sled support, take 2 (WIP)

### PR DESCRIPTION
**Pull request by jespersm**
Since I can no longer run the "bin" tests locally, I'll just try and build the PR for the Sled datastore, one piece at a time

### CODE REVIEW

**Comment by jespersm** [comment]
Now this should fix #79, and adds a way for the `bin` server to start an instance backed by Sled.

I'm still not able to run the "bin" tests locally, though, and I can't understand why (not even on master).

Benchmarks on my machine:
```
test rdb::normal_config::bench_bulk_insert                 ... bench: 117,911,097 ns/iter (+/- 6,793,321)
test rdb::normal_config::bench_create_edge                 ... bench:       9,776 ns/iter (+/- 2,459)
test rdb::normal_config::bench_create_vertex               ... bench:       4,665 ns/iter (+/- 1,271)
test rdb::normal_config::bench_get_edge_count              ... bench:       2,052 ns/iter (+/- 115)
test rdb::normal_config::bench_get_edges                   ... bench:       1,342 ns/iter (+/- 306)
test rdb::normal_config::bench_get_vertices                ... bench:       1,173 ns/iter (+/- 140)
test sledds::bench_bulk_insert                             ... bench: 298,968,269 ns/iter (+/- 13,294,011)
test sledds::bench_create_edge                             ... bench:      11,093 ns/iter (+/- 563)
test sledds::bench_create_vertex                           ... bench:       5,560 ns/iter (+/- 338)
test sledds::bench_get_edge_count                          ... bench:       1,722 ns/iter (+/- 99)
test sledds::bench_get_edges                               ... bench:       1,075 ns/iter (+/- 141)
test sledds::bench_get_vertices                            ... bench:         926 ns/iter (+/- 88)
```

**Comment by ysimonson** [comment]
When you run bin tests locally, does it stall out, or something else? I didn't manage to repro the issue locally unfortunately.

**Comment by jespersm** [comment]
Yes, it stalls while connecting.

**Comment by ysimonson** [comment]
Bin tests presently have a naive way of starting up servers, [where it allocates ports starting at 27616](https://github.com/indradb/indradb/blob/a319bdb35393fb117540fb91ce641f3a93f2d053/bin/src/common/tests.rs#L21-L22) - I wonder if that's causing problems. I'll work on a fix for that.

**Comment by ysimonson** [comment]
Merged in port randomization. That may fix the stalls, but this is all guesswork since I can't repro.

**Comment by jespersm** [comment]
I doubt that it would fix the stall problem, as what I'm seeing on macOS is related to how the client connects. It stalls before getting the connection through.

To experiment, I switched both server and client to using async_std::task::block_on to drive the tasks instead of using the LocalPool, and now it works fine for the async stuff (e.g. by making ClientTransaction::async_* public and calling them from within a. block_on), but obviously, you can't block inside the async context.

I'm no expert on async Rust yet, but I don't think you can mix blocking and async code like it's done in the ClientConnection.
I can see how it should work using the LocalPool, but it appears to stall for me. I'm guessing that the issue is some kind of incompatibility between asynd_std used for the sockets, and futures-executors.

(I can see how it's done in reqwest::blocking, and I think that would require everything to use the same executor, Tokio in their case)

**Comment by ysimonson** [comment]
My understanding is that the futures crate's executors are built to play nicely with other executors. It's also what the [cap'n proto examples do](https://github.com/capnproto/capnproto-rust/tree/279b765cb911de85c290ecb06d5c44a9ddfee971/capnp-rpc/examples/hello-world).

Mixing sync and async does seem risky though. As unconventional as that is, I'm hoping to keep the general structure since it makes it possible to reuse all the existing lib tests in bin.

Regardless, I'll keep digging.

**Comment by jespersm** [comment]
I just ran the cap'n'proto tests on my Mac, and they worked fine. Odd.

But the two issues aren't related, you should consider this Sled MR independently as the tests run fine on the CI machine. This issue could be related only to my machine, who knows.

**Comment by ysimonson** [comment]
Other than the bin tests stalling, is this PR at a good point? If so, I can just go ahead and take care of the last few bits before merging it in.

**Comment by jespersm** [comment]
The bin tests do not stall on the CI infra, only on my machine.

But master's bin tests also stall on my machine. I don't see the issues as connected.

File: lib/src/sledds/tests.rs [hunk 1]
```diff
@@ -0,0 +1 @@
+
```
Comment (ysimonson) [inline lib/src/sledds/tests.rs:1]: No need for this module

**Comment by ysimonson** [comment]
This looks great. There's a couple of very minor touch-ups, but I can address them after merge.

**Pull request merged**